sdk/eventhub/azure-eventhub/azure/eventhub/_buffered_producer/_buffered_producer.py

```diff
@@ -85,39 +85,39 @@ def put_events(self, events, timeout_time=None):
         # Put single event or EventDataBatch into the queue.
         # This method would raise OperationTimeout if the queue does not have enough space for the input and
         # flush cannot finish in timeout.
-        try:
-            new_events_len = len(events)
-        except TypeError:
-            new_events_len = 1
-        if self._max_buffer_len - self._cur_buffered_len < new_events_len:
-            _LOGGER.info(
-                "The buffer for partition %r is full. Attempting to flush before adding %r events.",
-                self.partition_id,
-                new_events_len,
-            )
-            # flush the buffer
-            with self._lock:
+        with self._lock:
```

Comment:
I wouldn't lock the whole method like this.
When flush is called on line 100, it will also try to take the lock and likely hang.
We only need to lock access to the queue.

Reply:
I also changed the behavior on the async side to match it.